File: gluecodium/src/test/resources/smoke/lambdas/output/android/com/example/smoke/Lambdas.java
/*
 *
 */
package com.example.smoke;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import com.example.NativeBase;
import java.util.List;
import java.util.Map;
public final class Lambdas extends NativeBase {
    /**
     * @hidden
     */
    static class ProducerImpl extends NativeBase implements Producer {
        protected ProducerImpl(final long nativeHandle, final Object dummy) {
            super(nativeHandle, new Disposer() {
                @Override
                public void disposeNative(long handle) {
                    disposeNativeHandle(handle);
                }
            });
        }
        private static native void disposeNativeHandle(long nativeHandle);
        @NonNull
        public native String apply();
    }
    /**
     * @hidden
     */
    static class ConfounderImpl extends NativeBase implements Confounder {
        protected ConfounderImpl(final long nativeHandle, final Object dummy) {
            super(nativeHandle, new Disposer() {
                @Override
                public void disposeNative(long handle) {
                    disposeNativeHandle(handle);
                }
            });
        }
        private static native void disposeNativeHandle(long nativeHandle);
        /**
         * <p>Should confuse everyone thoroughly
         * @param p0
         * @return
         */
        @NonNull
        public native Lambdas.Producer confuse(@NonNull final String p0);
    }
    /**
     * @hidden
     */
    static class ConsumerImpl extends NativeBase implements Consumer {
        protected ConsumerImpl(final long nativeHandle, final Object dummy) {
            super(nativeHandle, new Disposer() {
                @Override
                public void disposeNative(long handle) {
                    disposeNativeHandle(handle);
                }
            });
        }
        private static native void disposeNativeHandle(long nativeHandle);
        public native void apply(@NonNull final String p0);
    }
    /**
     * @hidden
     */
    static class IndexerImpl extends NativeBase implements Indexer {
        protected IndexerImpl(final long nativeHandle, final Object dummy) {
            super(nativeHandle, new Disposer() {
                @Override
                public void disposeNative(long handle) {
                    disposeNativeHandle(handle);
                }
            });
        }
        private static native void disposeNativeHandle(long nativeHandle);
        public native int apply(@NonNull final String p0, final float index);
    }
    /**
     * @hidden
     */
    static class NullableConfuserImpl extends NativeBase implements NullableConfuser {
        protected NullableConfuserImpl(final long nativeHandle, final Object dummy) {
            super(nativeHandle, new Disposer() {
                @Override
                public void disposeNative(long handle) {
                    disposeNativeHandle(handle);
                }
            });
        }
        private static native void disposeNativeHandle(long nativeHandle);
        @Nullable
        public native Lambdas.Producer apply(@Nullable final String p0);
    }
    @FunctionalInterface
    public interface Producer {
        @NonNull
        String apply();
    }
    /**
     * <p>Should confuse everyone thoroughly
     */
    @FunctionalInterface
    public interface Confounder {
        /**
         * <p>Should confuse everyone thoroughly
         * @param p0
         * @return
         */
        @NonNull
        Lambdas.Producer confuse(@NonNull final String p0);
    }
    @FunctionalInterface
    public interface Consumer {
        void apply(@NonNull final String p0);
    }
    @FunctionalInterface
    public interface Indexer {
        int apply(@NonNull final String p0, final float index);
    }
    @FunctionalInterface
    public interface NullableConfuser {
        @Nullable
        Lambdas.Producer apply(@Nullable final String p0);
    }
    /**
     * For internal use only.
     * @hidden
     * @param nativeHandle The SDK nativeHandle instance.
     * @param dummy The SDK dummy instance.
     */
    protected Lambdas(final long nativeHandle, final Object dummy) {
        super(nativeHandle, new Disposer() {
            @Override
            public void disposeNative(long handle) {
                disposeNativeHandle(handle);
            }
        });
    }
    private static native void disposeNativeHandle(long nativeHandle);
    @NonNull
    public native Lambdas.Producer deconfuse(@NonNull final String value, @NonNull final Lambdas.Confounder confuser);
    @NonNull
    public static native Map<Integer, String> fuse(@NonNull final List<String> items, @NonNull final Lambdas.Indexer callback);
}

File: gluecodium/src/test/resources/smoke/lambdas/output/android-kotlin/com/example/smoke/Lambdas.kt
/*

 *
 */

@file:JvmName("LambdasExtensions")


package com.example.smoke

import com.example.NativeBase

class Lambdas : NativeBase {

    fun interface Producer {


        fun apply() : String
    }

    /**
     * @suppress
     *
     * This class is used to represent C++ implementations of the interface or lambda in Kotlin.
     * It is instantiated by JNI and should not be used by the end users.
     */
    private class ProducerImpl : NativeBase, Producer {
        protected constructor(nativeHandle: Long, @Suppress("UNUSED_PARAMETER") tag: Any?)
            : super(nativeHandle, { disposeNativeHandle(it) }) {}



        override external fun apply() : String



        companion object {
            @JvmStatic private external fun disposeNativeHandle(nativeHandle: Long)
        }
    }
    /**
     * Should confuse everyone thoroughly
     */
    fun interface Confounder {
        /**
         * Should confuse everyone thoroughly
         * @param p0
         * @return
         */

        fun confuse(p0: String) : Lambdas.Producer
    }

    /**
     * @suppress
     *
     * This class is used to represent C++ implementations of the interface or lambda in Kotlin.
     * It is instantiated by JNI and should not be used by the end users.
     */
    private class ConfounderImpl : NativeBase, Confounder {
        protected constructor(nativeHandle: Long, @Suppress("UNUSED_PARAMETER") tag: Any?)
            : super(nativeHandle, { disposeNativeHandle(it) }) {}

        /**
         * Should confuse everyone thoroughly
         * @param p0
         * @return
         */

        override external fun confuse(p0: String) : Lambdas.Producer



        companion object {
            @JvmStatic private external fun disposeNativeHandle(nativeHandle: Long)
        }
    }
    fun interface Consumer {


        fun apply(p0: String) : Unit
    }

    /**
     * @suppress
     *
     * This class is used to represent C++ implementations of the interface or lambda in Kotlin.
     * It is instantiated by JNI and should not be used by the end users.
     */
    private class ConsumerImpl : NativeBase, Consumer {
        protected constructor(nativeHandle: Long, @Suppress("UNUSED_PARAMETER") tag: Any?)
            : super(nativeHandle, { disposeNativeHandle(it) }) {}



        override external fun apply(p0: String) : Unit



        companion object {
            @JvmStatic private external fun disposeNativeHandle(nativeHandle: Long)
        }
    }
    fun interface Indexer {


        fun apply(p0: String, idx: Float) : Int
    }

    /**
     * @suppress
     *
     * This class is used to represent C++ implementations of the interface or lambda in Kotlin.
     * It is instantiated by JNI and should not be used by the end users.
     */
    private class IndexerImpl : NativeBase, Indexer {
        protected constructor(nativeHandle: Long, @Suppress("UNUSED_PARAMETER") tag: Any?)
            : super(nativeHandle, { disposeNativeHandle(it) }) {}



        override external fun apply(p0: String, idx: Float) : Int



        companion object {
            @JvmStatic private external fun disposeNativeHandle(nativeHandle: Long)
        }
    }
    fun interface NullableConfuser {


        fun apply(p0: String?) : Lambdas.Producer?
    }

    /**
     * @suppress
     *
     * This class is used to represent C++ implementations of the interface or lambda in Kotlin.
     * It is instantiated by JNI and should not be used by the end users.
     */
    private class NullableConfuserImpl : NativeBase, NullableConfuser {
        protected constructor(nativeHandle: Long, @Suppress("UNUSED_PARAMETER") tag: Any?)
            : super(nativeHandle, { disposeNativeHandle(it) }) {}



        override external fun apply(p0: String?) : Lambdas.Producer?



        companion object {
            @JvmStatic private external fun disposeNativeHandle(nativeHandle: Long)
        }
    }


    /**
     * For internal use only.
     * @suppress
     * @param nativeHandle The handle to resources on C++ side.
     * @param tag Tag used by callers to avoid overload resolution problems.
     */
    protected constructor(nativeHandle: Long, @Suppress("UNUSED_PARAMETER") tag: Any?)
        : super(nativeHandle, { disposeNativeHandle(it) }) {}





    external fun deconfuse(value: String, confuser: Lambdas.Confounder) : Lambdas.Producer




    companion object {
        @JvmStatic private external fun disposeNativeHandle(nativeHandle: Long)


        @JvmStatic external fun fuse(items: List<String>, callback: Lambdas.Indexer) : Map<Int, String>
    }
}


